server/src/test/java/org/elasticsearch/snapshots/SnapshotsServiceTests.java

```diff
@@ -459,40 +624,89 @@ protected void assertSnapshotOrGenericThread() {
-            masterService.setClusterStatePublisher((clusterChangedEvent, publishListener, ackListener) -> {
-                final AtomicInteger applyCounter = new AtomicInteger(testClusterNodes.nodes.size());
-                testClusterNodes.nodes.values().forEach(
-                    n ->
-                        deterministicTaskQueue.scheduleNow(() -> {
-                            assertThat(n.currentState.get().version(), lessThan(clusterChangedEvent.state().version()));
-                            ClusterChangedEvent adjustedEvent = changeEventForNode(clusterChangedEvent, n.node);
-                            n.repositoriesService.applyClusterState(adjustedEvent);
-                            n.snapshotsService.applyClusterState(adjustedEvent);
-                            n.snapshotShardsService.clusterChanged(adjustedEvent);
-                            n.indicesClusterStateService.applyClusterState(adjustedEvent);
-                            n.currentState.set(adjustedEvent.state());
-                            if (applyCounter.decrementAndGet() == 0) {
-                                publishListener.onResponse(null);
-                                ackListener.onCommit(TimeValue.timeValueMillis(deterministicTaskQueue.getLatestDeferredExecutionTime()));
-                            }
-                        }));
-            });
-            masterService.setClusterStateSupplier(currentState::get);
-            masterService.start();
             ClusterState stateForNode = stateForNode(initialState, node);
-            currentState.set(stateForNode);
-            clusterService.getClusterApplierService().setInitialState(stateForNode);
+            final CoordinationState.PersistedState persistedState = new InMemoryPersistedState(0L, stateForNode);
+            coordinator = new Coordinator(node.getName(), clusterService.getSettings(),
+                clusterService.getClusterSettings(), transportService, namedWriteableRegistry,
+                allocationService, masterService, () -> persistedState,
+                hostsResolver -> testClusterNodes.nodes.values().stream().filter(n -> n.node.isMasterNode())
+                    .map(n -> n.node.getAddress()).collect(Collectors.toList()),
+                clusterService.getClusterApplierService(), random());
+            masterService.setClusterStatePublisher(coordinator);
+            coordinator.start();
+            masterService.start();
             clusterService.getClusterApplierService().setNodeConnectionsService(new NodeConnectionsService(clusterService.getSettings(),
                 deterministicTaskQueue.getThreadPool(), transportService));
             clusterService.getClusterApplierService().start();
             indicesService.start();
             indicesClusterStateService.start();
+            coordinator.startInitialJoin();
+        }
+
+        private class ConfigurableDisruptableMockTransport extends DisruptableMockTransport {
```

Comment:
This whole class is a 75% copy of what is done in CoordinatorTests and we can probably way dry things up here.